libtenzir/builtins/contexts/bloom_filter.cpp

```diff
@@ -0,0 +1,240 @@
+                             "invalid 'key' parameter; 'key' must be a string");
+    }
+    auto key_column = slice.schema().resolve_key_or_concept(*key_field);
+    if (not key_column) {
+      // If there's no key column then we cannot do much.
+      return update_result{record{}};
+    }
+    auto [key_type, key_array] = key_column->get(slice);
+    auto context_array = std::static_pointer_cast<arrow::Array>(
+      to_record_batch(slice)->ToStructArray().ValueOrDie());
+    auto key_values = values(key_type, *key_array);
+    auto key_values_list = list{};
+    auto context_values = values(slice.schema(), *context_array);
+    for (const auto& value : key_values) {
+      auto materialized_key = materialize(value);
+      bloom_filter_.add(materialized_key);
+      key_values_list.emplace_back(std::move(materialized_key));
+    }
+    auto query_f = [key_values_list = std::move(key_values_list)](
+                     parameter_map params) -> caf::expected<expression> {
+      auto column = params["field"];
+      if (not column) {
+        return caf::make_error(ec::invalid_argument,
+                               "missing 'field' parameter");
+      }
+      return expression{
+        predicate{
+          field_extractor(*column),
+          relational_operator::in,
+          data{key_values_list},
+        },
+      };
+    };
+    return update_result{.update_info = show(),
+                         .make_query = std::move(query_f)};
+  }
+
+  auto update(chunk_ptr, context::parameter_map)
+    -> caf::expected<update_result> override {
+    // TODO: We're getting chunks piece-meal currently and therefore cannot
+    // determine the boundary of the stream. In theory, we should accumulate the
+    // chunks until the upstream pipeline is complete. What we need in the
+    // context API is a signal that we're done.
+    return ec::unimplemented;
+  }
+
+  auto update(context::parameter_map) -> caf::expected<update_result> override {
+    return caf::make_error(ec::unimplemented,
+                           "bloom-filter context can not be updated with void");
+  }
+
+  auto save() const -> caf::expected<chunk_ptr> override {
+    std::vector<std::byte> buffer;
+    if (auto err = convert(bloom_filter_, buffer))
+      return add_context(err, "failed to serialize Bloom filter context");
+    return chunk::make(std::move(buffer));
+  }
+
+private:
+  dcso_bloom_filter bloom_filter_;
+};
+
+class plugin : public virtual context_plugin {
+  auto initialize(const record&, const record&) -> caf::error override {
+    return caf::none;
+  }
+
+  auto name() const -> std::string override {
+    return "bloom-filter";
+  }
+
+  auto make_context(context::parameter_map parameters) const
+    -> caf::expected<std::unique_ptr<context>> override {
+    auto capacity = parameters["capacity"];
+    if (not capacity)
+      return caf::make_error(ec::parse_error, "no --capacity provided");
+    auto fpp = parameters["fp-prob"];
```

Comment:
Looking at other bloom-filter implementations, it seems to me that this option is most commonly called `fp-probability`. I recommend switching to that.